server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -814,6 +840,198 @@ private Scheduler.Cancellable fetchClusterFormationInfo(
+             * this poll attempt is the current one. If they are not equal, that means that
+             * cancelPollingRemoteStableMasterHealthIndicatorService() has been called on this poll attempt but this thread is not yet
+             * aware. So we cancel the Cancellable in cancellableReference if it is not null. Note that
+             * remoteStableMasterHealthIndicatorTask can be null.
+             */
+            if (cancellableReference.equals(remoteStableMasterHealthIndicatorTask)) {
+                /*
+                 * Because this is not synchronized with the cancelPollingRemoteStableMasterHealthIndicatorService() method, there is a
+                 * slim chance that we will add a task here for a poll that has already been cancelled. But when it completes and runs
+                 * rescheduleDiagnosticsFetchConsumer() we will then see that remoteStableMasterHealthIndicatorTask does not equal
+                 * cancellableReference, so it will not be run again.
+                 */
+                try {
+                    DiscoveryNode masterEligibleNode = getMasterEligibleNodes().stream().findAny().orElse(null);
+                    cancellableReference.set(
+                        fetchCoordinationDiagnostics(
+                            masterEligibleNode,
+                            responseConsumer.andThen(rescheduleDiagnosticsFetchConsumer(responseConsumer, cancellableReference))
+                        )
+                    );
+                } catch (EsRejectedExecutionException e) {
+                    if (e.isExecutorShutdown()) {
+                        logger.trace("Not rescheduling request for cluster coordination info because this node is being shutdown", e);
+                    } else {
+                        throw e;
+                    }
+                }
+            } else {
+                Scheduler.Cancellable cancellable = cancellableReference.get();
+                if (cancellable != null) {
+                    cancellable.cancel();
+                }
+            }
+        };
+    }
+
+    /**
+     * This method returns quickly, but in the background schedules to query the remote masterEligibleNode's cluster diagnostics in 10
+     * seconds unless cancel() is called on the Cancellable that this method returns.
+     * @param masterEligibleNode The masterEligibleNode to poll for cluster diagnostics. This masterEligibleNode can be null in the case
+     *                           when there are not yet any master-eligible nodes known to this masterEligibleNode's PeerFinder.
+     * @param responseConsumer The consumer of the cluster diagnostics for the masterEligibleNode, or the exception encountered while
+     *                         contacting it
+     * @return A Cancellable for the task that is scheduled to fetch cluster diagnostics
+     */
+    private Scheduler.Cancellable fetchCoordinationDiagnostics(
```

Comment:
This is very similar to `fetchClusterFormationInfo` ? Could we extract a common method to be used in both cases - this common method would additionaly have parameters for the action name, request and a response reader?

Reply:
Done.

Reply:
❤️